osidb/models.py

```diff
@@ -1807,6 +1841,54 @@ class Meta:
             "external_system_id",
             "created_dt",
         )
+        constraints = [
+            models.UniqueConstraint(
+                fields=["flaw", "order"], name="unique_per_flaw_comment_nums"
+            ),
+        ]
+
+    def bzsync(self, *args, bz_api_key=None, **kwargs):
+        """
+        Bugzilla sync of the FlawComment instance and of the related Flaw instance.
+        """
+
+        self.save()
+
+        # Comments need to be synced through flaw
+        # If external_system_id is blank, BugzillaSaver posts the new comment
+        # and FlawCollector loads the new comment and updates this FlawComment
+        # instance to match bugzilla.
+        self.flaw.save(*args, bz_api_key=bz_api_key, **kwargs)
+
+    def _validate_comment_order(self):
+        """
+        Check that this comment's order (comment number) is directly following
+        the last existing comment of the given flaw. This checks that the
+        client has a sufficiently up-to-date idea of the flaw when submitting
+        the comment.
+
+        Usually, a flaw has a comment number 0 as its description. However,
+        if the description is missing, the first non-description comment must
+        start with number 1.
+
+        This validation is skipped for comments that have external_system_id,
+        which bugzilla always provides, so it's used just for client-submitted
+        new comments.
+        """
+        if self.flaw.comments.count() == 0:
```

Comment:
I wonder whether the count can be actually zero. If you `self.flaw` exists then that flaw has at least one comment which is `self`. Maybe you need to exclude comments without the `external_system_id`?

Reply:
I think I saw flaws that had zero comments. It might be just in tests, but in principle it's probably possible. Without this condition, 53 tests fail. Maybe caused by uncommitted instances? This is what I tried to run with:

```
        (...)
        This validation is skipped for comments that have external_system_id,
        which bugzilla always provides, so it's used just for client-submitted
        new comments.
        """
        proper_following_order = (
            1 + self.flaw.comments.aggregate(Max("order"))["order__max"]
        )
        if self.external_system_id == "" and proper_following_order != self.order:
            raise ValidationError(
        (...)
```

Reply:
It is definitely possible that the new functionality breaks existing tests and often it is just because they become wrong in the new context. Failing tests do not have to mean that your change is wrong.

It can happen that there is a flaw with no comments. My confusion however is how could it happen when there is an existing comment (`self`) and its `flaw` attribute is an existing flaw. Then it seems logical to me that this exact flaw has at least one comment (`self`) and thus I do not understand when the number could be zero.

Reply:
I agree with Ondrej.

Reply:
Resolved in commit 76a8581033d95a8a77a598c876457e2b162a158f

Reply:
Also found this gem https://github.com/bugzilla/bugzilla/blob/2f203a360eab075742d42fd228a4233a3d45d221/Bugzilla/Bug.pm#L3760

Reply:
Detective work :smile: